Refactor: Optimize Packet Handling in api/tunnel.go

HynoR opened the pull request:
a memory optimization patch for packet handling in the `api/tunnel.go`. The changes aim to reduce memory allocations and improve performance when reading and writing packets. Below are the key updates grouped by theme:

* Buffer Pool Implementation
 Using Sync.Pool for managing reusable byte slices and initialized global buffers for packet buffer pooling. This reduces the need for frequent memory allocations.

* Refactor `TunnelDevice` interface for the function `ReadPacket` (May have conflict)
Every `ReadPacket` will create a new memory allocation for its buf, it can be reduced by put it in input parameter instead of new output
This Changes may need to be review and evaluate

*  Update Wireguard-go to latest version (Testing)
Ref: https://github.com/Diniboy1123/usque/issues/17
Need Testing.

Diniboy1123 reviewed (commented):
Hey,

This is super exciting, can't wait to merge this. Really clever small trick to avoid re-allocations. I only really found small nitpicks that I would like you to resolve before merging this. I wrote individual comments on those parts.

Next time I would prefer if you would split version bumps to a different PR as the rest could have been merged already, but I will just wait now until the version fix is rebased and then merge hopefully.

Thanks for the hard work spent on profiling, it seems to be better based on my quick tests as well. Will experiment more tomorrow.

Let me know what you think about those comments!

HynoR commented on api/tunnel.go:
```diff
@@ -6,18 +6,62 @@ import (
+		panic("capacity must be greater than 0")
+	}
+	if capacity < 1280 {
+		capacity = 1280
```
Removed the limit of buffer size, I still have no idea on the best buffer size, so this code is just copy from the MTU flags setting, But It may cause some problem when the size is too small.

HynoR commented on api/tunnel.go:
```diff
@@ -27,16 +71,27 @@ type NetstackAdapter struct {
+	packetBufs := tunnelBufPool.Get().([][]byte)
+	sizes := tunnelSizesPool.Get().([]int)
+
+	defer func() {
```
You are right, leave something in global area is a 'hack' code, So I have moved these buffers into NetStack Struct.

HynoR commented:
Ready for review!
#### About Buffer Size:
Removed the limit of buffer size, I still have no idea on the best buffer size, and I only found the recommend size in Stackoverflow that is 2048, so this code is just copy from the MTU flags setting, But It may cause some problem when the size is too small. 
#### About Global Buffer
Moved into NetstackAdapter struct.
#### Go Mod
I have run `go mod tidy` and committed

HynoR commented:
I rolled back dependency version to original, maybe I should not update it in this pr. We can create another pr to update wg-go and compare the improvement

HynoR commented on api/tunnel.go:
```diff
@@ -35,27 +37,26 @@ func (n *NetBuffer) Put(buf []byte) {
 		buf: sync.Pool{
 			New: func() interface{} {
-				return make([]byte, capacity)
+				b := make([]byte, capacity)
```
no need to add pointer to refactor it

HynoR reviewed (commented):
it Looks good.

HynoR commented:
your commit is good, we can test if it is better than the previous. Also, if using a pointer instead of the original slice in []byte sync.Pool buffer can reduce memory allocate, that might explain why the `New` Function in go-sock5  use a lot heap memory when cocurrent.

Diniboy1123 commented:
Personally I wouldn't want to change the signature of Get and Put to use pointers. I was thinking about the same thing yesterday, whether we could spare some allocations with that too, but my conclusion was that `[]byte` should be a reference type. If you pass that instead of a pointer and dereference it , you will end up doing no copies. The header is probably copied though so there is that, but that shouldn't be too significant. And I think working with `[]byte` is generally friendlier from an API perspective than working with pointers.

Though it's not a bad idea to confirm this theory of mine with a debugger like [`dlv`](https://github.com/go-delve/delve) to see what's actually going on when we pass by value.

>  we can test if it is better than the previous.

I took a look on the heap `pprof` profile and there seems to be less allocations now.

There is also something I am not completely happy about. The idea with the `api` package  is that its reusable by 3rd party apps. This implies that all function calls can be safely called in parallel. Right now we have a global:

```go
var packetBufferPool *NetBuffer
```

Which is not ideal once you have multiple goroutines from the same 3rd party app opening many tunnels. So one last refactor could be that, to create a new `NetBuffer` pool inside `MaintainTunnel` and then we should be good to go.

> Plus, we can use squash merge to hide and clean up my useless commit.

I am personally not a big fan of squashing as it masks the history which could help other developers. Especially since your change isn't just a minor one, commit history makes sense. I just didn't want to merge the undone commits and the empty merge request. But I should have force pushed on your previous changes, so those should be gone.

HynoR commented:
Ok, to be honest I am not very familiar with the buffer code — most of my changes were based on examples from others. actually I'm an ops engineer with an interest in Go, not a Go expert, so I'm still learning and thanks a lot for your review and corrections!
And I am now deployed this branch on my server and still testing to see if it need to be improved.

Diniboy1123 commented:
No worries, I am also no Go expert, just a random university student. Also had to look up some stuff in the golang source quick to be confident about my responses. 😉 I explicitely chose Go because of the quic-go, connect-ip-go and gvisor netstack triad. If that wouldn't exist, we would probably have a rust codebase by now.

Anyway I think your change makes sense overall and should increase performance quite a bit, so thanks for that once again. Keep testing, let me know how it performs and if everything goes well, we can merge this by the afternoon (CEST).
